### PR TITLE
cmd/snap-bootstrap: changes to be able to boot classic rootfs

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
If the model reveals that this is a classic system:

- Do not mount ubuntu-seed
- Do not mount base snap
- Do not mount snapd snap on first boot, we'll use initially the one
  in the deb

This also needs changes in the initramfs to mount /sysroot differently
depending on whether this is UC or classic, and to avoid starting
populate-writable for the latter.

### CODE REVIEW

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 1]
```diff
@@ -1617,7 +1635,11 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
 		return err
 	}
 
-	typs := []snap.Type{snap.TypeBase, snap.TypeGadget, snap.TypeKernel}
+	// order in the list must not change as it determines the mount order
+	typs := []snap.Type{snap.TypeGadget, snap.TypeKernel}
+	if !isClassic {
+		typs = append([]snap.Type{snap.TypeBase}, typs...)
+	}
```
Comment (mardy) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1642]: This is a bit confusing, because AFAIK we don't have gadget and kernel snaps on classic, do we? So, I would expect them also to be under the `!isClassic` condition.

Comment (alfonsosanchezbeato) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1642]: We will have them 🙂

**Review by mardy — APPROVED** [review]
LGTM, thanks!

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 2]
```diff
@@ -1493,47 +1493,63 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
...
-	//            need it to be writable for i.e. transitioning to recover mode
+	// The model is now measured, use it to check if this is a classic install
+	model, err := mst.UnverifiedBootModel()
+	if err != nil {
+		return err
+	}
+	isClassic := model.Classic()
+	if isClassic {
+		logger.Noticef("generating mounts for classic system, run mode")
+		boot.InitramfsWritableDir = boot.InitramfsDataDir
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1512]: this needs to be done somehow differently, redefining globals in other packages unless they are hooks is not something we do

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 3]
```diff
@@ -1493,47 +1493,63 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
...
+	// 2. mount ubuntu-seed if non-classic system
 	systemdOpts := &systemdMountOptions{
 		NeedsFsck: true,
 		Private:   true,
 	}
-	if err := doSystemdMount(fmt.Sprintf("/dev/disk/by-partuuid/%s", partUUID), boot.InitramfsUbuntuSeedDir, systemdOpts); err != nil {
-		return err
+	if !isClassic {
+		// use the disk we mounted ubuntu-boot from as a reference to find
+		// ubuntu-seed and mount it
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1524]: as we discussed medium term we sometimes will have ubuntu-seed, the correct thing might be to try to find it? and then later when we mount gadget double check if we expected it or not, and error consequently

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 4]
```diff
@@ -1546,20 +1562,22 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
...
 		NeedsFsck: true,
-		NoSuid:    true,
-		Private:   true,
+	}
+	if !isClassic {
+		// fsck and mount with nosuid to prevent snaps from being able to bypass
+		// the sandbox by creating suid root files there and trying to escape the
+		// sandbox
+		dataMountOpts.NoSuid = true
+		dataMountOpts.Private = true
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1573]: this doesn't comment why we also don't want Private true

**Review by pedronis — CHANGES_REQUESTED** [review]
did a pass, it needs some refinements and clarifications

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: Changed now

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1585]: I've added a comment. I don't use private here because I think that in general the behavior the user sees should be as close as possible to a regular classic system.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: Changed, I pass around an argument where needed instead.

**Comment by alfonsosanchezbeato** [comment]
@pedronis thanks for the review, I've addressed your comments now.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 5]
```diff
@@ -1480,89 +1482,117 @@ func maybeMountSave(disk disks.Disk, rootdir string, encrypted bool, mountOpts *
...
 	// use the disk we mounted ubuntu-boot from as a reference to find
 	// ubuntu-seed and mount it
 	partUUID, err := disk.FindMatchingPartitionUUIDWithFsLabel("ubuntu-seed")
 	if err != nil {
-		return err
+		if isClassic {
+			// If there is no ubuntu-seed on classic, that's fine
+			if _, ok := err.(disks.PartitionNotFoundError); !ok {
+				return "", err
+			}
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1534]: I don't see code below to check that the expectation from the gadget about this is met

File: boot/flags_test.go [hunk 6]
```diff
@@ -107,7 +107,7 @@ func (s *bootFlagsSuite) TestInitramfsActiveBootFlagsUC20InstallModeHappy(c *C)
 
 	setupRealGrub(c, blDir, "EFI/ubuntu", &bootloader.Options{Role: bootloader.RoleRecovery})
 
-	flags, err := boot.InitramfsActiveBootFlags(boot.ModeInstall)
+	flags, err := boot.InitramfsActiveBootFlags(boot.ModeInstall, boot.InitramfsWritableDir)
```
Comment (pedronis) [inline boot/flags_test.go:110]: should at least some of this tests use a directory value that wasn't the original one?

File: boot/initramfs_test.go [hunk 7]
```diff
@@ -620,10 +620,10 @@ func (s *initramfsSuite) TestInitramfsRunModeSelectSnapsToMount(c *C) {
 			c.Assert(err, IsNil, comment)
 
 			if t.expRebootPanic != "" {
-				f := func() { boot.InitramfsRunModeSelectSnapsToMount(t.typs, m) }
+				f := func() { boot.InitramfsRunModeSelectSnapsToMount(t.typs, m, boot.InitramfsWritableDir) }
```
Comment (pedronis) [inline boot/initramfs_test.go:623]: should this test use some other temp directory instead

**Review by pedronis — COMMENTED** [review]
thanks, some comments/questions

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1534]: Not sure if I understand this comment, the gadget that we mount comes from the ubuntu-boot partition, not from ubuntu-seed.

Comment (pedronis) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1534]: The gadget that we mount should come from the data partition, no?  What I'm seeing I don't see a check that we don't end up not having found a ubuntu-seed while our gadget volumes tell there should be one

File: boot/flags_test.go (hunk 6 above)
Comment (alfonsosanchezbeato) [reply boot/flags_test.go:110]: Added  tests

File: boot/initramfs_test.go (hunk 7 above)
Comment (alfonsosanchezbeato) [reply boot/initramfs_test.go]: Tests added too

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1536]: Good point, I've added checks for this

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the review, I've added changes now

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 8]
```diff
@@ -1676,7 +1679,34 @@ func generateMountsModeRun(mst *initramfsMountsState) (string, error) {
...
+			snapTypeToMountDir[snap.TypeGadget])
+		gadgetInfo, err := gadget.ReadInfo(gadgetDir, model)
+		if err != nil {
+			return "", err
+		}
+		seedDefinedInGadget := false
+	volLoop:
+		for _, vol := range gadgetInfo.Volumes {
+			for _, part := range vol.Structure {
+				if part.Role == gadget.SystemSeed &&
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1694]: checking role is enough I think, we check the part Name when we validate the gadget I think

**Review by pedronis — APPROVED** [review]
thanks. small remark

**Comment by pedronis** [comment]
this need a review also from @mvo5

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 9]
```diff
@@ -151,7 +154,7 @@ func generateInitramfsMounts() (err error) {
 	case "factory-reset":
 		err = generateMountsModeFactoryReset(mst)
 	case "run":
-		err = generateMountsModeRun(mst)
+		rootfsDir, err = generateMountsModeRun(mst)
```
Comment (mvo5) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:157]: Nice!

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 10]
```diff
@@ -1601,59 +1634,89 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
...
 			if err := doSystemdMount(snapPath, filepath.Join(boot.InitramfsRunMntDir, dir), mountReadOnlyOptions); err != nil {
-				return err
+				return "", err
+			}
+		}
+	}
+
+	// 4.4 check if we expected a ubuntu-seed partition from the gadget data
+	if isClassic {
+		gadgetDir := filepath.Join(boot.InitramfsRunMntDir,
```
Comment (mvo5) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1684]: (nitpick^9) same comment about linebreaks as above ;)

File: cmd/snap-bootstrap/main.go [hunk 11]
```diff
@@ -66,6 +66,9 @@ func parseArgs(args []string) error {
 	p := parser()
 
 	_, err := p.ParseArgs(args)
+	if err != nil {
+		logger.Noticef("execution error: %v", err)
```
Comment (mvo5) [inline cmd/snap-bootstrap/main.go:70]: Silly question - why is this added/needed? Our default logger logs to os.Stderr iirc and on error parseArgs will also fmt.Fprintf() to os.Stderr and exit 1 ? What am I missing :)

**Review by mvo5 — APPROVED** [review]
Thank you. This looks very nice. Tiny nitpicking inline and a question but nothing that blocks this PR.

File: cmd/snap-bootstrap/main.go (hunk 11 above)
Comment (alfonsosanchezbeato) [reply cmd/snap-bootstrap/main.go:70]: Great question in fact ;). I added this because otherwise you would not see the error in the console. Writing to os.stderr is not enough, you need to write to the journal so journald can output when configured to do so.

File: boot/bootstate20.go [hunk 12]
```diff
@@ -385,9 +385,10 @@ func (ks20 *bootState20Kernel) setNext(next snap.PlaceInfo, bootCtx NextBootCont
 // Choosing to boot/mount the base snap needs to be committed to the
 // modeenv, but no state needs to be committed when choosing to mount a
 // kernel snap.
-func (ks20 *bootState20Kernel) selectAndCommitSnapInitramfsMount(modeenv *Modeenv) (sn snap.PlaceInfo, err error) {
+func (ks20 *bootState20Kernel) selectAndCommitSnapInitramfsMount(modeenv *Modeenv, rootfsDir string) (sn snap.PlaceInfo, err error) {
 	// first do the generic choice of which snap to use
-	first, second, err := genericInitramfsSelectSnap(ks20, modeenv, TryingStatus, "kernel")
+	first, second, err :=
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply boot/bootstate20.go]: I've changed this now. Actually, it is not that I stick to 80 chars per line, I break lines when I consider a line too long to be easily read. It is about readibility and eye movement - newspaper and research articles use columns for a reason. Anyway, changed as the general style in snapd is what you say.

**Comment by alfonsosanchezbeato** [comment]
Rebased as there was a conflict with master
